alibi_detect/utils/missing_optional_dependency.py

```diff
@@ -0,0 +1,118 @@
+    + "the correct optional dependencies, run `pip install alibi-detect[$missing_dependency]` "
+    + "from the command line. For more information, check the Installation documentation "
+    + "at https://docs.seldon.io/projects/alibi-detect/en/latest/overview/getting_started.html."
+))
+
+
+ERROR_TYPES = {
+    "fbprophet": 'prophet',
+    "holidays": 'prophet',
+    "pystan": 'prophet',
+    "tensorflow_probability": 'tensorflow',
+    "tensorflow": 'tensorflow',
+    "torch": 'torch',
+    "pytorch": 'torch'
+}
+
+
+class MissingDependency:
+    """Missing Dependency Class
+    Used to replace any object that requires unmet optional dependencies. Attribute access or calling the __call__
+    method on this object will raise an error.
+    """
+    def __init__(self,
+                 object_name: str,
+                 err: Union[ModuleNotFoundError, ImportError],
+                 missing_dependency: str = 'all',):
+        """ Metaclass for MissingDependency classes
+        Parameters
+        ----------
+        object_name
+            Name of object we are replacing
+        missing_dependency
+            Name of missing dependency required for object
+        err
+            Error to be raised when the class is initialized or used
+        """
+        self.missing_dependency = missing_dependency
+        self.object_name = object_name
+        self.err = err
+
+    @property
+    def err_msg(self):
+        """Generate error message informing user to install missing dependencies."""
+        return err_msg_template.substitute(
+            object_name=self.object_name,
+            missing_dependency=self.missing_dependency)
+
+    def __getattr__(self, key):
+        """Raise an error when attributes are accessed."""
+        raise ImportError(self.err_msg) from self.err
+
+    def __call__(self, *args, **kwargs):
+        """If called, raise an error."""
+        raise ImportError(self.err_msg) from self.err
+
+
+def import_optional(module_name: str, names: Optional[List[str]] = None) -> Any:
+    """Import a module that depends on optional dependencies
+    Note: This function is used to import modules that depend on optional dependencies. Because it mirrors the python
+    import functionality its return type has to be `Any`. Using objects imported with this function can lead to
+    misspecification of types as `Any` when the developer intended to be more restrictive.
+    Parameters
+    ----------
+        module_name
+            The module to import
+        names
+            The names to import from the module. If None, all names are imported.
+    Returns
+    -------
+        The module or named objects within the modules if names is not None. If the import fails due to a
+        ModuleNotFoundError or ImportError then the requested module or named objects are replaced with instances of
+        the MissingDependency class above.
+    """
```

Comment:
Nitpick: I don't think numpy style docstrings give you indentend parameter lists (e.g. compare with `MissingDependency' class above)